### PR TITLE
WidgetTester.drag with time duration

### PR DESCRIPTION
**Pull request by CareF**
## Description

Add `duration` and `frequency` optional argument to implement operation that lasts for a period of time, used for performance test purposes to capture animation during an operation. 
Documents will be updated after #63390 lands. 

## Related Issues

#38329

## Tests

I added the following tests:

`picture_cache_perf__e2e_summary` in devicelab, as part of #62171 to test the new feature as well as to test the related performance. 

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_test/lib/src/controller.dart [hunk 1]
```diff
@@ -500,8 +508,9 @@ abstract class WidgetController {
     int buttons = kPrimaryButton,
     double touchSlopX = kDragSlopDefault,
     double touchSlopY = kDragSlopDefault,
+    Duration duration,
+    double frequency = 60.0,
   }) {
-    assert(kDragSlopDefault > kTouchSlop);
```
Comment (CareF) [inline packages/flutter_test/lib/src/controller.dart:504]: This one is removed due to duplicate of the assert in `dragFrom`.

File: packages/flutter_test/lib/src/controller.dart [hunk 2]
```diff
@@ -527,12 +536,94 @@ abstract class WidgetController {
...
```
Comment (CareF) [inline packages/flutter_test/lib/src/controller.dart:531]: I think assert the value that's going to be used is more meaningful than assert the global constant.

File: packages/flutter_test/lib/src/controller.dart [hunk 3]
```diff
@@ -477,6 +477,14 @@ abstract class WidgetController {
   /// system identifies the gesture as a fling, consider using [fling] instead.
   ///
   /// {@template flutter.flutter_test.drag}
+  /// If `duration` is not `null`, this is a convenience wrap of
+  /// [handlePointerEventRecord], with the input lasts for `duration` at the
+  /// given `frequency` (Hz).
```
Comment (liyuqian) [inline packages/flutter_test/lib/src/controller.dart:482]: Nit: consider reusing the documentations from `FlutterDriver` such as
```
[duration] specifies the length of the action. 

The move events are generated at a given [frequency] in Hz (or events per second). It defaults to 60Hz.

If [duration] is null, input events are assumed to happen at once, and [frequency] is ignored.

See also [LiveTestWidgetsFlutterBindingFramePolicy.benchmarkLive] for more accurate time control.
```

File: packages/flutter_test/lib/src/controller.dart [hunk 4]
```diff
@@ -527,12 +536,94 @@ abstract class WidgetController {
     int buttons = kPrimaryButton,
     double touchSlopX = kDragSlopDefault,
     double touchSlopY = kDragSlopDefault,
+    Duration duration,
+    double frequency = 60.0,
   }) {
-    assert(kDragSlopDefault > kTouchSlop);
+    assert(touchSlopX > kTouchSlop);
+    assert(touchSlopY > kTouchSlop);
```
Comment (liyuqian) [inline packages/flutter_test/lib/src/controller.dart:543]: The `touchSlop` logic seems to be much more sophisticated than the `scroll` logic in `flutter_driver`, and it could generate more input events than `duration * frequency`. Shall we just create a separate `drag` function that doesn't have `touchSlop`, but use `duration` and `frequency` to decide how many events to be generated?

Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart:543]: `touchSlop` comes in as a bug fix in #26438. I'm feeling it strange too but I don't have good enough reason to change it. I'm expecting in the duration-frequency one because `offset/(duration * frequency)` is very small (if the frequency and duration is reasonable), it will always go to single event for each individual input. However if a user is abusing these numbers (like duration * frequency gives 1), I don't want to trigger that bug, so I'm using this as a fall back.

File: packages/flutter_test/lib/src/controller.dart (hunk 3 above)
Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: Done

File: packages/flutter_test/lib/src/controller.dart (hunk 4 above)
Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart:545]: As for driver, I feel it's just we are being lucky not having gone into that bug.

Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: Double checked  #26438, without this `touchSlop` the widgets that are having issue is the cupertino style switcher. I guess scrolling is not when there's problem. In `flutter_driver`, there isn't a `drag` but always scrolling. My understanding is `flutter_driver` has not reached the stage of being able to do all operations and no one has used `flutter_driver` to test dragging a cupertino switch. However, changing the timed version to `scroll` doesn't sounds good to me because the operation is not always scrolling either (like the one in `macrobenchmarks/picture_cache`).

Comment (liyuqian) [reply packages/flutter_test/lib/src/controller.dart]: I'm a little inclined to keep it simple until we've seen actual abuses (like `duration * frequency` gives 1) that require us to add such complex logic. Otherwise, we risk of making many of our tests potentially hard to reason about without actually helping anything. Specifically, since we're going to use this in places where `FlutterDriver.scroll` was used, I felt just matching `FlutterDriver.scroll` input events would be good enough to not trigger any new bugs.

Additionally, the old `touchSlop` logic seems to be specifically designed for `duration == null` case and breaks a single move event into 2 move events. Using such logic to break n events into 2n events looks unnecessary and unintended (suppose the offset is really large so each event's offset is large enough to be separated).

Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: So I moved the timed part to `slowDrag` and `slowDragFrom`. PTAL

Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: I also added a unit test for `timedDrag`. One more thing that's worth mentioning (already in a comment), the PointerUpEvent and its corresponding PointerData doesn't provide `button` field.

File: packages/flutter_test/lib/src/controller.dart [hunk 5]
```diff
@@ -530,9 +535,25 @@ abstract class WidgetController {
   }) {
     assert(kDragSlopDefault > kTouchSlop);
     return TestAsyncUtils.guard<void>(() async {
-      final TestGesture gesture = await startGesture(startLocation, pointer: pointer, buttons: buttons);
+      final TestGesture gesture = await startGesture(
```
Comment (liyuqian) [inline packages/flutter_test/lib/src/controller.dart:538]: We can restore all the changes here as `_separateDragOffset` is not going to be used by anyone else

File: packages/flutter_test/lib/src/controller.dart [hunk 6]
```diff
@@ -577,27 +598,139 @@ abstract class WidgetController {
...
+      pointer: pointer,
+      buttons: buttons,
+      frequency: frequency,
+    );
+  }
+
+  /// Attempts a series of [PointerEvent]s to simulate a drag operation in the
+  /// `duration` time.
+  ///
+  /// If the middle of the widget is not exposed, this might send
```
Comment (liyuqian) [inline packages/flutter_test/lib/src/controller.dart:661]: There's no widget given so "the middle of the widget" seems to be an irrelevant documentation.

File: packages/flutter_test/lib/src/controller.dart [hunk 7]
```diff
@@ -577,27 +598,139 @@ abstract class WidgetController {
...
+  /// events to another object.
+  ///
+  /// This is the timed version of [drag]. This may or may not result in a
```
Comment (liyuqian) [inline packages/flutter_test/lib/src/controller.dart:664]: `[drag]` -> `[dragFrom]`? I suggest putting the identical comments in a template, so it's clear what can be shared and what cannot.

File: packages/flutter_test/test/controller_test.dart [hunk 8]
```diff
@@ -550,6 +550,43 @@ void main() {
     },
   );
 
+  testWidgets(
+    'WidgetTester.slowDrag must respect buttons',
```
Comment (liyuqian) [inline packages/flutter_test/test/controller_test.dart:554]: `slowDrag` -> `timedDrag`

File: packages/flutter_test/lib/src/controller.dart (hunk 7 above)
Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: Done.

File: packages/flutter_test/test/controller_test.dart (hunk 8 above)
Comment (CareF) [reply packages/flutter_test/test/controller_test.dart]: Done.

File: packages/flutter_test/lib/src/controller.dart (hunk 6 above)
Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: Removed.

File: packages/flutter_test/lib/src/controller.dart (hunk 5 above)
Comment (CareF) [reply packages/flutter_test/lib/src/controller.dart]: Done.

**Review by liyuqian — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
fly-by comment: There's a merge conflict, so the automerge via "waiting for tree to be green" will fail.
